### PR TITLE
Seek multiple times and support Buffers

### PR DESCRIPTION
**Pull request by vweevers**
This PR enables multiple calls to `iterator.seek(target)` while iterating, and adds support for Buffer keys. The primary use case is [skip scanning](https://oracle-base.com/articles/9i/index-skip-scanning).

``` js
iter.next(function(err,key,value){
  if (key.toString() === 'a') {
    iter.seek('g')
    iter.next(etc)
  }
})
```

I renamed the variable `key` to `target`, because it better expresses the fact that it's not an exact match.

Like before, `seek()` respects the iterator direction (reverse or not), which differs from the native LevelDB seek behavior. The native seek lands on or _after_ a target, but in leveldown, if reverse is true, it lands on or _before_ the target (at least it should, breaking test cases are welcome).

It also accounts for keys in the cache. If one of those matches the target, no native seek is performed. Instead, the cache is truncated so that the following `next()` yields the target key.

Note I'm not a C++ guy, and this PR is bigger than I intended. But what I basically did, is move the initial seek in `GetIterator()` to a separate private function (`IteratorSeek`) so that both `GetIterator` (for the regular initial seek) and `Seek` can use it.

### CODE REVIEW

**Comment by vweevers** [comment]
Note that in the current implementation (from PR #185), a `seek()` to a non-existent key yields undefined in the following `next()` call (i.e. it's an exact match, unlike the native seek). I'm not sure if that was the intended behavior. /cc @mafintosh ?

**Comment by mafintosh** [comment]
@vweevers ah no. i think the native way is better.

**Comment by vweevers** [comment]
That the [build is failing](https://travis-ci.org/Level/leveldown/jobs/109181758) on io.js seems unrelated to this PR?

**Comment by juliangruber** [comment]
i reran the build, now it passed. this is the 2nd random failure this week, i'll open an issue to investigate that.

File: test/iterator-test.js [hunk 1]
```diff
@@ -35,35 +58,250 @@ make('iterator is seekable', function (db, t, done) {
...
+})
+
+make('iterator seek stays in range', function(db, t, done) {
+  db.batch(pairs(10), function(err){
+    t.error(err, 'no error from batch()')
+
+    var pending = 0
+
+    expect({ gt: 5 }, 5, 6) // hm, should it?
+    expect({ gt: 5 }, 6, 6)
```
Comment (juliangruber) [inline test/iterator-test.js]: what's that about?

Comment (vweevers) [inline test/iterator-test.js]: Given that there's a sequence of numbers in the db, what should this do?

```js
var ite = db.iterator({gt: '5'})
ite.seek('5')
ite.next(function(err, key, value){
  // ?
})
```

- throw, because the seek target (5) is out of range
- yield `undefined`
- yield 6 (because even though the target is out of range, the seek lands on 6, which makes it valid) (this is the current behavior)

File: src/iterator.cc [hunk 2]
```diff
@@ -205,34 +206,26 @@ void checkEndCallback (Iterator* iterator) {
...
-    if (iterator->reverse) {
-      dbIterator->SeekToLast();
-    } else {
-      dbIterator->SeekToFirst();
-    }
     if (dbIterator->Valid()) {
-      int cmp = dbIterator->key().compare(*key);
-      if (cmp > 0 && iterator->reverse) {
-        dbIterator->SeekToFirst();
+      if (target->compare(dbIterator->key().ToString()) < 0) {
```
Comment (vweevers) [inline src/iterator.cc]: I'm currently refactoring things so that:

- this bit can be moved to `Iterator::IteratorSeek`, keep all the comparisons there
- and to not perform the actual seek right away, but at the next read

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,6 @@
 node_modules/
 build/
+Release/
```
Comment (rvagg) [inline .gitignore]: what's making this directory?

File: iterator.js [hunk 4]
```diff
@@ -14,11 +16,32 @@ function Iterator (db, options) {
...
+
+  // Check if target is cached
+  var length = this.cache ? this.cache.length : 0
+
+  if (length) {
+    if (typeof target === 'string') {
+      target = new Buffer(target)
+    }
+
+    for(var i=length-1; i>0; i-=2) {
```
Comment (rvagg) [inline iterator.js]: whitespace all the things please `for (var i = length - 1; i > 0; i -= 2) {`

File: iterator.js [hunk 5]
```diff
@@ -14,11 +16,32 @@ function Iterator (db, options) {
...
+  if (this._nexting)
+    throw new Error('cannot call seek() before next() has completed')
+  if (typeof target !== 'string' && !Buffer.isBuffer(target))
+    throw new Error('seek() requires a string or buffer target')
+
+  // Check if target is cached
+  var length = this.cache ? this.cache.length : 0
+
+  if (length) {
+    if (typeof target === 'string') {
```
Comment (rvagg) [inline iterator.js]: remove parens over simple single `if`

File: test/iterator-test.js (hunk 1 above)
Comment (juliangruber) [inline test/iterator-test.js]: yielding 6 makes most sense to me. this way, the iterator constructor could be implemented like

```js
function iterator(opts){
  if (opts.gt) this.seek(opts.gt)
}
```

so my vote is to keep it the way leveldb implements it

File: iterator.js [hunk 6]
```diff
@@ -14,11 +16,32 @@ function Iterator (db, options) {
...
+      target = new Buffer(target)
+    }
+
+    for(var i=length-1; i>0; i-=2) {
+      var c = compare(this.cache[i], target)
+      if (this.reverse ? c <= 0 : c >= 0) {
+        return void this.cache.splice(i+1, length)
```
Comment (rvagg) [inline iterator.js]: whitespace, `i + `

also, `return void`? a bit too clever, probably needs multiple lines to achieve this

File: .gitignore (hunk 3 above)
Comment (juliangruber) [inline .gitignore]: apparently that's from windows builds

File: iterator.js (hunk 6 above)
Comment (juliangruber) [inline iterator.js]: agreed

File: .gitignore (hunk 3 above)
Comment (vweevers) [inline .gitignore]: gyp on windows (msbuild)

**Comment by rvagg** [comment]
The implicit assumption here is that `ltgt` performs exactly the same way as the native comparator, I don't think this is a safe assumption and IMO any comparison of keys should be performed by the native comparator rather than pulling in `ltgt` which better belongs at the levelup and up layer.

If the current `seek()` implementation does this:
> to a non-existent key yields undefined in the following `next()` call 
then that does seem like a bug to me, although I was basically absent for that PR so perhaps there was something intentional here I'm missing?

**Comment by vweevers** [comment]
@rvagg the problem is that the iterator might already be finished (with the cache filled) when one calls `seek()`. But I agree with you. What I could try, is to reset the state (`iterator.finished` in the JS world), empty the cache, then do the seek.

**Comment by vweevers** [comment]
Good, that works. Don't know why I didn't think of that sooner. I'll remove `ltgt` :)

File: test/iterator-test.js (hunk 1 above)
Comment (vweevers) [inline test/iterator-test.js]: Actually, I was conflating two things. The above example lands on 5 natively, it is leveldown who then skips to 6 (because of [this line](https://github.com/Level/leveldown/blob/master/src/iterator.cc#L112) - which only applies to `gt`). But that's just a single `Next()` call, which means that a seek to 4 on `{ gt: 5 }` yields undefined: leveldown sees that 4 doesn't satify the range, does a `Next()`, now at 5, but still not satisfied, so it ends the iterator.

That the native seek lands on or after a target, is a useful property for seeking *to a range*, but not for seeking *within a range*. It also makes little sense for consumers to seek outside of a range that they themselves defined. E.g. why seek to 10 if you specified `{ lt: 5 }` as range. So I propose to guard against this. If a target is out of range, don't do the seek, just yield undefined. It's much easier to reason about.

Comment (juliangruber) [inline test/iterator-test.js]: | E.g. why seek to 10 if you specified { lt: 5 } as range. 

that makes total sense, I missed that

File: test/iterator-test.js [hunk 7]
```diff
@@ -3,31 +3,38 @@ const test       = require('tape')
...
-make('iterator throws if key is not a string', function (db, t, done) {
-  var ite = db.iterator()
-  var error
-  try {
-    ite.seek()
-  } catch (e) {
-    error = e
-  }
+make('iterator seek throws if key is not a string or buffer', function (db, t, done) {
+  var keys = [void 0, 10, true, null]
```
Comment (juliangruber) [inline test/iterator-test.js]: is there a reason not to use `undefined` instead of `void 0`? void is quite uncommon these days

File: test/iterator-test.js [hunk 8]
```diff
@@ -3,31 +3,38 @@ const test       = require('tape')
...
+  var pending = keys.length
 
-  t.ok(error, 'had error')
-  t.end()
+  keys.forEach(function(key){
+    var ite = db.iterator(), error;
```
Comment (juliangruber) [inline test/iterator-test.js]: iirc we're not mixing assignments and declarations on the same line

File: test/iterator-test.js [hunk 9]
```diff
@@ -35,35 +42,280 @@ make('iterator is seekable', function (db, t, done) {
   })
 })
 
-make('reverse seek in the middle', function (db, t, done) {
+make('iterator is seekable with buffer', function (db, t, done) {
+  var ite = db.iterator()
+  ite.seek(new Buffer('two'))
```
Comment (juliangruber) [inline test/iterator-test.js]: you can omit the `new` from Buffer creation

File: test/iterator-test.js [hunk 10]
```diff
@@ -35,35 +42,280 @@ make('iterator is seekable', function (db, t, done) {
...
+  ite.next(function (err, key, value) {
+    t.error(err, 'no error from next()')
+    t.same(key.toString(), 'two', 'key matches')
```
Comment (juliangruber) [inline test/iterator-test.js]: `t.same` uses `==` whereas `t.equal` uses `===`. i think we should use `t.equal` where possible

File: test/iterator-test.js (hunk 8 above)
Comment (vweevers) [inline test/iterator-test.js]: is there a specific code style (document) I can refer to?

Comment (juliangruber) [inline test/iterator-test.js]: not really, the code really is the code style guide

File: test/iterator-test.js (hunk 10 above)
Comment (ralphtheninja) [inline test/iterator-test.js]: Afaik `t.same` is an alias for `t.deepEqual`

Comment (juliangruber) [inline test/iterator-test.js]: that's true as well. but that's not required here

Comment (vweevers) [inline test/iterator-test.js]: true, I'll change it to `t.equal`, and also replace the alias `t.is` (in other tests) with `t.equal` for consistency. Is it ok to also update the original test ("iterator is seekable") to use `t.equal`?

Comment (juliangruber) [inline test/iterator-test.js]: i think that should be a separate pull request, to make this one easier to merge

**Comment by vweevers** [comment]
Thanks for all the comments! I undid some changes to variable names and tests, to make the PR cleaner (see diff [with whitespace ignored](https://github.com/Level/leveldown/pull/256/files?diff=split&w=1)). And I refactored the seeking:

- It now respects the range: `seek('5')` on `{ gt: '5' }` yields undefined
- The seek is not performed right away, but on the next read. Question about that: is there risk of accidental GC?

File: src/iterator.cc [hunk 11]
```diff
@@ -205,38 +221,34 @@ void checkEndCallback (Iterator* iterator) {
 
 NAN_METHOD(Iterator::Seek) {
   Iterator* iterator = Nan::ObjectWrap::Unwrap<Iterator>(info.This());
-  iterator->GetIterator();
-  leveldb::Iterator* dbIterator = iterator->dbIterator;
-  Nan::Utf8String key(info[0]);
 
-  dbIterator->Seek(*key);
-  iterator->seeking = true;
+  // TODO: check if slice size > 0?
```
Comment (vweevers) [inline src/iterator.cc]: There are [comments](https://github.com/vweevers/leveldown/blob/iterator-seek/src/iterator.cc#L383) saying "a Slice can't have length 0". Can someone shine some light on that? What should happen on `seek("")`?

File: src/iterator.cc [hunk 12]
```diff
@@ -76,65 +77,80 @@ Iterator::~Iterator () {
...
+  bool isFirstCall = GetIterator();
+
+  // skip read if manual seek is out of range
+  if (outOfRange)
+    return false;
+
+  // do manual seek, initial seek, or move to next item
+  if (target != NULL) {
+    IteratorSeek(target);
+    target = NULL;
```
Comment (vweevers) [inline src/iterator.cc]: Does this suffice, or can it cause a leak?

**Comment by vweevers** [comment]
Except for squashing the commits (which I'll do after your code review), I think this PR is done. Notable changes include:

- `seek('')` or `seek(Buffer(0))` throws. In my own tests, seeking to an empty Slice behaved like a `SeekToFirst()`, but after reading reports about segfaults (Level/levelup#112) and LevelDB being unable to handle zero-length Slices (#31), I decided the safer route is to throw. If someone needs a public `seekToFirst/Last()` that's something for a future PR.
- the Slice created on `seek(key)` is properly disposed (using `DisposeStringOrBufferFromSlice`) when `NextWorker` ends (i.e. after seeking), when the `Iterator` is destroyed, or when `seek(key)` is called in succession (even though that wouldn't make sense)
- I simplified the post-seek key comparisons in reverse mode. Full test suite passes but if someone would like to verify my thought process, I wrote it up [in a gist](https://gist.github.com/vweevers/390db48442ef8b0139ff).

**Comment by ralphtheninja** [comment]
Cool. I haven't had the time to go through this yet. Hopefully get some time on the weekend.

**Comment by peakji** [comment]
Any news on this? Skip scanning is really useful.

**Comment by vweevers** [comment]
@peakji this PR is too big to review, I now think. If we want to make any progress, it should be split up. I haven't had the time to do so.

**Comment by vweevers** [comment]
There are multiple features here; each deserves their own PR:

- Buffer support
- Lazy seeking
- Multiple seeks (by resetting `finished` state)
- Land on or before target
- Merge the logic of initial and manual seek (but IIRC there was dead code so first we might want to ensure that the current C++ has full code coverage, see this [lcov example](https://github.com/nodejs/node-addon-examples/pull/43));
- Don't seek outside `ltgt` range.
